[Serve] [Docs] Add consolidated `Model Composition` user guide

shrekris-anyscale opened the pull request:
Signed-off-by: Shreyas Krishnaswamy <shrekris@anyscale.com>

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

This change adds introductory deployment graph documentation.

Links to updated documentation:
* [Model Composition](https://ray--26860.org.readthedocs.build/en/26860/serve/model_composition.html)
* [Examples Overview](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/index.html)
* [Deployment Graph Pattern Overview](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns.html)
  * [Pattern: Linear Pipeline](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns/linear_pipeline.html)
  * [Pattern: Branching Input](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns/branching_input.html)
  * [Pattern: Conditional](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns/conditional.html)

## Related issue number

Closes #27443 and closes #27470.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - The documentation code has been added to the `doc_code` directory and has assertions to catch regressions.

simon-mo commented:
Please let us know when it's ready for review.

architkulkarni commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,219 @@
+
+Ray Serve provides the `DAGDriver`, which routes HTTP requests through your call graph. As mentioned in [the call graph section](deployment-graph-intro-call-graph), the `DAGDriver` has one required argument: the `FunctionNode` or `MethodNode` representing your call graph's final output.
+
+The `DAGDriver` also has one more optional keyword argument: `http_adapter`. `http_adapters` are functions that get run on the HTTP request before it's passed into the graph. Ray Serve provides a handful of these adapters, so you can rely on them to conveniently handle the HTTP parsing while focusing your attention on the graph itself.
```
Would be nice to link to the HTTP adapters section

architkulkarni commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,219 @@
+
+The `DAGDriver` also has one more optional keyword argument: `http_adapter`. `http_adapters` are functions that get run on the HTTP request before it's passed into the graph. Ray Serve provides a handful of these adapters, so you can rely on them to conveniently handle the HTTP parsing while focusing your attention on the graph itself.
+
+For instance, we can use the Ray Serve-provided `json_request` adapter to simplify our [arithmetic call graph](deployment-graph-intro-arithmetic-graph) by eliminating the `unpack_request` function. Here's the revised call graph and driver:
```
Nice, it might be good to add a tip box linking to this section somewhere in the call graph section.  Saying something like "if you don't want to do this unpacking manually, jump to [this section]"

shrekris-anyscale replied:
Good idea, I added a tip.

architkulkarni commented on doc/source/serve/deployment-graph/http_endpoint_for_dag_graph.md:
`````diff
@@ -12,7 +12,7 @@ This example shows how to configure ingress component of the deployment graph, s
 
 ````{note}
-1. Serve provide a special driver ([DAGDriver](deployment-graph-e2e-tutorial.md)) to accept the http request and drive the dag graph execution
+1. Serve provide a special driver ([DAGDriver](deployment_graph_e2e_tutorial.md)) to accept the http request and drive the dag graph execution
`````
```suggestion
1. Serve provides a special driver ([DAGDriver](deployment_graph_e2e_tutorial.md)) to accept the HTTP request and drive the dag graph execution
```

architkulkarni commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,219 @@
+
+* compose your Ray Serve deployments together with the **Deployment Graph** API
+* serve your applications that use multi-model inference, ensemble models, ML model composition, or mixed business logic/model inference workloads
+* independently scale each of your ML models and business logic steps
```
I didn't see this covered, maybe we just need a brief note about num_replicas somewhere?

shrekris-anyscale replied:
Good catch, I added a note.

architkulkarni commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,219 @@
+
+The rest of the call graph uses another `FunctionNode` and `MethodNode` to finish the chain of arithmetic. `add_2_output` is bound to the `subtract_one_fn` deployment, producing the `subtract_1_output` `FunctionNode`. Then, the `subtract_1_output` is bound to the `add_3.add` method, producing the `add_3_output` `MethodNode`. This `add_3_output` `MethodNode` represents the final output from our ch ...
+
+To run the call graph, you need to use a driver. Drivers are deployments that process the call graph that you've written and route incoming requests through your deployments based on that graph. Ray Serve provides a driver called `DAGDriver` used on line 38:
```
Is there a more general Driver section or API reference we can link to here?

shrekris-anyscale replied:
No, I don't think one exists yet.

architkulkarni commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
````diff
@@ -0,0 +1,219 @@
+# Intro to Deployment Graphs
+
+```{note}
+Note: This feature is in Alpha, so APIs are subject to change.
````
Which parts are Alpha?  For example is `.bind()` alpha? It's in the Getting Started docs, which suggests it's GA

architkulkarni reviewed (commented):
Looks great!  At a high level, I could see users being a bit confused about:
- when to use ClassNode vs when to use MethodNode, what are the advantages/disadvantages? Maybe you could spell it out a bit more
- how to decide whether to use the Deployment Graph API (which is alpha) or to use the non-deployment graph API, which is GA
If we can clarify those I think this is ready to go.  I also left some more minor comments

shrekris-anyscale commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,219 @@
+* serve your applications that use multi-model inference, ensemble models, ML model composition, or mixed business logic/model inference workloads
+* independently scale each of your ML models and business logic steps
+
+Ray Serve's **Deployment Graph** API lets you compose your deployments together by describing how to route a request through your deployments. This is particularly useful if you're using ML model composition or mixing business logic and model inference in your application. You can encapsulate each of your models and each of your business logic steps in independent deployments. Then, you can chain ...
```
Hmm, yeah this is a bit tricky. Even single-node deployments are using the deployment graph API (such as `bind` and `serve.run`), so users should always be using this API. The rest of the documentation doesn't really dwell on this.

It might be nice to acknowledge this here, though. Do you have any suggestions on how we do that?

architkulkarni replied:
In that case I think we can just leave it as is!

dmatrix replied:
I wonder if you want to have subheadings for the landing page that reads:
What are Deployments Graphs?
Why and when to use Deployment Graphs?
How to use them?

The developer is looking for this quick information what, why or when, and how? I have suggested that format Ray AIR. The how bit becomes your guickstart.

shrekris-anyscale commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,219 @@
+The basic building block of deployment graphs is the `DeploymentNode`. There are
+three types of `DeploymentNodes`:
+
+* `ClassNode`: a `DeploymentNode` containing a Python class bound to its constructor's arguments
```
Good catch, I was alluding to `.bind()`. I made it more explicit.

shrekris-anyscale commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
````diff
@@ -0,0 +1,219 @@
+```
+
+(deployment-graph-intro-call-graph)=
+## The Call Graph: MethodNodes and FunctionNodes
````
Yeah they are– these are the nodes that make up the "Call graph" since they actually specify what functions and methods get called.

dmatrix replied:
Again, do you want to use present tenses here than future tenses? For example, _this call starts a Ray cluster_.
This is always true.

If you are grammar snob then you can argue an action _will trigger_ another action shortly in the future, hence should be a future tense.

It's like saying "this medication will help you relax" vs "this medication helps you relax." Since the latter is always true, present tense is more effective and expressive, less laborious to read. Just a thought.

shrekris-anyscale commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
````diff
@@ -0,0 +1,219 @@
+deployment_graph = DAGDriver.bind(add_3_output)
+```
+
+The `DAGDriver` needs to be bound to the `FunctionNode` or `MethodNode` representing the final output of our graph. This `bind` call returns a `ClassNode` that you can run in `serve.run` or `serve run`. Running this `ClassNode` will also deploy the rest of the graph's deployments.
````
Actually a `DAGDriver` can be bound to a `ClassNode`. I've added a note.

shrekris-anyscale commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
````diff
@@ -0,0 +1,219 @@
+# Intro to Deployment Graphs
+
+```{note}
+Note: This feature is in Alpha, so APIs are subject to change.
````
Good question– most of these parts should actually be beta soon, so I can update the note once that change goes through. In general, this API is the one that users should use.

shrekris-anyscale commented:
> when to use ClassNode vs when to use MethodNode, what are the advantages/disadvantages? Maybe you could spell it out a bit more

@architkulkarni users don't really choose to between `ClassNode` and `MethodNode`. If they write a call graph, they'll usually need `MethodNodes`. Are you anticipating users being confused about when to use a call graph?

architkulkarni commented:
For example, the EchoClass is a ClassNode with a `__call__` function that could be bound to the `http_request` inside a call graph, right?  What's the difference between that and a MethodNode for a `.echo()` method that does the same thing?  When would you use a ClassNode and not a MethodNode?

shrekris-anyscale commented:
> For example, the EchoClass is a ClassNode with a __call__ function that could be bound to the http_request inside a call graph, right?

Do you mean bind it like this?

```python
foo_node = EchoClass("foo")
with InputNode() as http_request:
    foo_output = foo_node.__call__.remote(http_request)

serve.run(DAGDriver.bind(foo_output))
```

I think if you `bind` the `__call__` function like that, `foo_output` is still a `MethodNode`, so functionally, there's no difference than binding the `echo` method.

architkulkarni commented:
Ah makes sense, I think I was confusing myself with constructor arguments vs method arguments.

architkulkarni reviewed (approved):
Looks good, just need to resolve merge conflicts and ask codeowners for approval

edoakes reviewed (approved):
Looks really great, thanks @shrekris-anyscale and thanks for the careful review @architkulkarni.

Let's merge this one as it stands so you are unblocked to restructure the pages and collapse into a single "Model Composition" user guide. I will do a deeper copy-edit pass next week once we've restructured.

simon-mo commented:
@dmatrix or @sriram-anyscale  can you please approve or suggest changes for copy-writing pass?

dmatrix commented:
Will take a look at it shortly

dmatrix commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
````diff
@@ -0,0 +1,270 @@
+```
+
+This section should help you:
+
````
This section should help you to:

dmatrix commented on doc/source/serve/deployment-graph/deployment_graph_intro.md:
```diff
@@ -0,0 +1,270 @@
+* `MethodNode`: a `DeploymentNode` representing a `ClassNode`'s method bound to arguments (using `bind`) that will be used to invoke the method
+* `FunctionNode`: a `DeploymentNode` containing a Python function bound to arguments (using `bind`) that will be used to invoke the function
+
+The next two sections will discuss how to construct and connect these nodes to form deployment graphs.
```
Use present tense, because it's true all the time and is in active voice too. Such future tense is expected in academic papers, but for developer docs, keep them conversational if possible. 

"The next two sections discuss how to..."

dmatrix reviewed (commented):
Can one of you send me the read-doc...html generated from this PR? I want to see the code and flow. 

I have left some comments about tenses and perhaps structure of what, why & when, and how (leading to a quick start).

dmatrix commented:
This landing page for the user guide is too sparse in text. We may want to think about how a developer will want to navigate and what information he or she is looking for.  We don't need two headings back to back about deployment graphs.

One way to improve the flow and readability is add contextual information. Just few sentences here and there. Here is the new page I suggest:


### Introduction to Deployment Graphs
Ray Serve’s Deployment Graph offers simple API to compose your serve deployments together by declaring how to route a request through your various deployments strung together. This is particularly useful if you’re using ML model composition or mixing business logic and model inference in your application. Encapsulating each of your models and each of your business logic steps in independent deployments gives you added flexibility to update or scale individual deployments. Chained together, these deployments comprise a single unit as deployment graph, a direct acyclic graph. 

To understand key concepts and what nodes comprise as a deployment graph, follow this [guide](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/deployment_graph_intro.html#serve-intro-to-deployment-graphs). 

### An end-to-end walk throughs
A typical production serve deployment does not exist in isolation. Normally, it's a collection of deployments. To see some of these composite of end-to-end deployments, follow the tutorials below.

- End to end [tutorials](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/deployment_graph_e2e_tutorial.html) for deployment graphs

### Common Deployment Patterns
There are various common deployments patterns used in production, which you can compose using  deployment graphs. Below are a few examples how to use them:

[Chain nodes with same class and different args](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/chain_nodes_same_class_different_args.html)

[Combine two nodes with passing same input in parallel](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/combine_two_nodes_with_passing_input_parallel.html)

[Control flow based on user inputs](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/control_flow_based_on_user_inputs.html)

[Visualize DAG during development](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/visualize_dag_during_development.html)

[Http endpoint for dag graph](https://ray--26860.org.readthedocs.build/en/26860/serve/deployment-graph/http_endpoint_for_dag_graph.html)







![Screen Shot 2022-08-05 at 6 43 00 AM](https://user-images.githubusercontent.com/1117597/183090310-e07906af-fad0-40b6-8061-3eb0137d2903.png)

shrekris-anyscale commented:
Thanks for your detailed suggestions @dmatrix! I've updated the intro documentation to use only present tense, and I incorporated your suggestions into the landing page.

> The developer is looking for this quick information what, why or when, and how? I have suggested that format Ray AIR. The how bit becomes your guickstart.

I think this makes a lot of sense. I've restructured the documentation to talk about model composition in general (and discuss deployment graphs as a part of that). This makes the documentation address a "job-to-be-done" instead of listing Ray Serve jargon. Do you mind skimming the updated "Model Composition: Deployment Graphs" page to see if the structure makes sense?

dmatrix commented on doc/source/serve/deployment-graph.md:
```diff
@@ -2,13 +2,20 @@
+(serve-deployment-graph-patterns)=
+## Common Deployment Graph Patterns
+
+There are various common deployments patterns used models in production. These patterns allow you to compose a single deployment graph. Below are a few examples how to construct them:
```
patterns used with models in production

edoakes commented on doc/source/serve/deployment-graph/model_composition.md:
```diff
@@ -0,0 +1,273 @@
+(serve-model-composition)=
+
+# Model Composition: Deployment Graphs
```
```suggestion
# Model Composition
```

edoakes commented on doc/source/serve/deployment-graph/model_composition.md:
````diff
@@ -0,0 +1,273 @@
+# Model Composition: Deployment Graphs
+
+```{note}
+Note: This feature is in Alpha, so APIs are subject to change.
````
this should only apply to call graphs.

edoakes commented on doc/source/serve/deployment-graph/model_composition.md:
```diff
@@ -0,0 +1,273 @@
+
+Ray Serve's **deployment graph** API lets you compose your deployments together by describing how to route a request through your deployments. This is particularly useful if you're using ML model composition or mixing business logic and model inference in your application. You can encapsulate each of your models and each of your business logic steps in independent deployments. Then, you can conne ...
+
+## Implicit Composition: Cross-Deployment Calls
```
I don't think users will understand what "implicit" vs "explicit" means in this context. let's just "call it what it is" instead:
```suggestion
## Calling Deployments using ServeHandle
```

edoakes replied:
this more closely matches my changes to user guide as well

edoakes commented on doc/source/serve/deployment-graph/model_composition.md:
````diff
@@ -0,0 +1,273 @@
+```
+
+(deployment-graph-intro-call-graph)=
+## Explicit Composition: The Call Graph
````
```suggestion
## Deployment Graph API
```

edoakes commented on doc/source/serve/deployment-graph.md:
```diff
@@ -2,13 +2,20 @@
+## Introduction to Deployment Graphs
 
-- [E2E Tutorials](./deployment-graph/deployment-graph-e2e-tutorial.md)
+Ray Serve’s Deployment Graph offers a simple API to compose your Ray Serve deployments together by declaring how to route a request through them. This is particularly useful when using ML model composition or mixing business logic and model inference in your application. Encapsulating each of your models and each of your business logic steps in independent deployments gives you added flexibility  ...
```
@shrekris-anyscale in general let's only try to expose complexity to users when it's necessary. In this section, I don't think we should introduce the idea of a "graph" at all until we get to the actual DAG API section.

For example, a more incremental flow would be like this:
- This intro says it's going to walk you through how to accomplish model composition (focus on job-to-be-done!), particularly how to compose multiple deployments containing ML models or business logic.
- Then the first section introduces having deployments call into each other (no need to call this a graph at all yet, the user doesn't need to be exposed to the complexity).
- Then the final section introduces the full call graph/DAG API.

edoakes commented on doc/source/serve/deployment-graph/model_composition.md:
```diff
@@ -0,0 +1,273 @@
+
+You can connect your models together in an implicit deployment graph by calling deployment methods from within other deployments. This lets you divide your application's steps (such as preprocessing, model inference, and post-processing) into independent deployments that can be independently scaled and configured. You can route requests to a single "driver" deployment that can partially process a ...
+
+### ClassNodes
```
I don't think we need to introduce ClassNode at all yet. The user only needs to think about deployments and handles in order to grok this section and accomplish basic model composition.

shrekris-anyscale replied:
Good call, I moved the subsection about `ClassNodes` into the next section (about the `Deployment Graph API`).

edoakes commented:
@shrekris-anyscale see the key concepts page for how I'm discussing the basic multi-model composition https://github.com/ray-project/ray/pull/27565

edoakes commented:
@shrekris-anyscale FYI you have a merge conflict.

edoakes commented:
@shrekris-anyscale looks like the code sample isn't rendering ?

<img width="843" alt="Screen Shot 2022-08-08 at 11 40 52 AM" src="https://user-images.githubusercontent.com/9871461/183469242-f061a618-b173-4189-b505-fb47cba65f22.png">

edoakes commented on doc/source/serve/tutorials/deployment-graph-patterns/conditional.md:
````diff
@@ -0,0 +1,72 @@
+```
+
+:::{note}
+`combine` takes in intermediate values from the call graph as the individual arguments, `value1` and `value2`. You can also aggregate and pass these intermediate values as a list argument. However, this list will contain references the values, rather than the values themselves. You must explicitly use `ray.get` to get the actual values before using them.
````
not that in a deployment using `ray.get` is an antipattern because it blocks the `asyncio` loop. we should be recommending to `await `instead. We should make this abundantly clear in the model composition user guide. Could you add a short subsection on it?

shrekris-anyscale replied:
I added a warning about this.

shrekris-anyscale commented:
I've updated the PR description with new links:

* [Model Composition](https://ray--26860.org.readthedocs.build/en/26860/serve/model_composition.html)
* [Examples Overview](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/index.html)
* [Deployment Graph Pattern Overview](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns.html)
  * [Pattern: Linear Pipeline](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns/linear_pipeline.html)
  * [Pattern: Branching Input](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns/branching_input.html)
  * [Pattern: Conditional](https://ray--26860.org.readthedocs.build/en/26860/serve/tutorials/deployment-graph-patterns/conditional.html)

shrekris-anyscale commented:
I'm not sure how to add the `Deployment Graph Patterns` overview page to the `Examples` overview page:

<img width="631" alt="Screen Shot 2022-08-08 at 11 00 16 AM" src="https://user-images.githubusercontent.com/92341594/183483331-ca8c57ff-0685-4b0b-b95a-c126512bd6e0.png">

The issue is that if I add the `Deployment Graph Patterns` overview page directly to the `Examples` page, the `Deployment Graph Patterns` won't appear as a drop-down menu on the sidebar.

Instead, if I add the patterns overview to the `_toc.yml` file, it appears in the sidebar as a drop-down menu, but it doesn't appear on the `Examples` overview page itself.

<img width="310" alt="Screen Shot 2022-08-08 at 11 04 38 AM" src="https://user-images.githubusercontent.com/92341594/183483991-be78e2b9-d8c0-48d1-88a5-91e37bcbba4f.png">

If I add it to both, it appears twice on the sidebar, both times as drop-down menus.

I could add an explicit link under the `Other topics` list (which only contains `Deploying Ray Serve` at the moment), but I'm not sure if that's the best approach.

edoakes commented:
@shrekris-anyscale the content is looking great! Let's have @dmatrix do one more copy-editing pass and let me know when the code sample is updated to properly `await` the handle.

shrekris-anyscale commented:
@edoakes I've updated the code sample to use `await`. The content is ready for a review.

edoakes commented:
@shrekris-anyscale lint failure
